implement Database abstract class and SqlDatabase subclass with basic…

maxwilliams94 opened the pull request:
… tests. Basic implementation of a cli (not finished or tested)

Structure of project also fixed, __init__ usage updated as well as notes, todos, test todos files created.

Commit:
implement Database abstract class and SqlDatabase subclass with basic tests. Basic implementation of a cli (not finished or tested)

Structure of project also fixed, __init__ usage updated as well as notes, todos, test todos files created.

DomWilliams0 commented on pydo/__init__.py:
```diff
@@ -0,0 +1,4 @@
+__version__ = 0.1
```
Should use strings for versions instead of floats, otherwise one day you'll have something like `v0.30000001`

DomWilliams0 commented on pydo/config/config.py:
```diff
@@ -0,0 +1,28 @@
+
+
+def load_config(path=None):
+    if path is None:
```
What if path is not `None`? Looks like lines 9-13 should be unindented (and cf_path renamed to `path`)?

DomWilliams0 commented on pydo/config/config.py:
```diff
@@ -0,0 +1,28 @@
+            with open(cf_path, 'r') as f:
+                return json.load(f)
+        except FileNotFoundError:
+            raise FileNotFoundError
```
There's no point in using try/except if you just reraise the exception

DomWilliams0 commented on pydo/database/database.py:
```diff
@@ -0,0 +1,42 @@
+        self.connection = None
+        self.complete_path = None
+        if config_path is None:
+            self.config_path = default_cf_path()
```
This is a prime place for a ternary 
```suggestion
            self.config_path = default_cf_path() if config_path is None else config_path
```

DomWilliams0 commented on pydo/database/database_factory.py:
```diff
@@ -0,0 +1,9 @@
+class DatabaseFactory:
+    @staticmethod
+    def database(db_type, db_name, db_path) -> Database:
+        if db_type == "sqlite":
```
Should have an `else` and throw an exception. You could also use an enum instead of a string for `db_type`

DomWilliams0 commented on pydo/database/sqlite.py:
```diff
@@ -0,0 +1,53 @@
+        if self.name.endswith(SqliteDatabase.db_ext):
+            self.complete_path = path.join(self.path, self.name)
+        else:
+            self.complete_path = path.join(self.path,
```
Minor but there's a lot of duplication across this if/else. And in py3 it's best to use pathlib instead of string manipulation for paths, e.g. https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.with_suffix

DomWilliams0 commented on pydo/database/sqlite.py:
```diff
@@ -0,0 +1,53 @@
+
+    def initialise(self):
+        self.load()  # Connecting to non-existent db file creates a new db file
+        self.cursor.execute("""CREATE TABLE tasks (
```
For when you have more big static SQL strings, it'd be worth putting them all into a separate module

DomWilliams0 commented on pydo/database/sqlite.py:
```diff
@@ -0,0 +1,53 @@
+        self.connection.commit()
+
+    def close(self):
+        self.connection.close()
```
Worth setting `self.connection` to None here?

DomWilliams0 commented on tests/test_sql.py:
```diff
@@ -0,0 +1,37 @@
+                                                db_path='.')
+        db.initialise()
+        db_full_path = os.path.join('.', 'test_sql.db')
+        assert os.path.exists(db_full_path)
```
You should use a proper test runner for tests, which will involve using the `self.assertXXX` functions instead of the `assert` keyword: https://docs.python.org/3/library/unittest.html

DomWilliams0 reviewed (changes requested):
gg

Commit:
- thematic changes to Database constructors
- correct indentation for when a specific config file is loaded
- self.connection and self.cursor are None'd when the database is closed
